### ISSUE TITLE
termdebug: switch to tokenized messages

### ISSUE THREAD

**Issue by GitMensch** [vim/vim#10300]
Currently everything is "anonymous and asynchronous" in termdebug; we send an instruction and do a best guess when evaluating the response what the response is about. This issue is about tracking a change to this (whenever it may be done)

[ ] remove message anonymity by adding tokens to the messages sent (which is in general the suggested and commonly documented way)
[ ] for evaluation: use the token to _know_ what the response is for

This also allows to fix the guessing about the stop state:

* sent out an instruction that will tell us if the inferior is running or not
* internally block _all_ further messages that potentially tinker with the run state (easiest: all messages, possibly also block interaction with the GDB buffer)
* that way wait for the answer to the message we sent out (has its token on return so can be distinguished)
* handle the answer - we now know the current state
* unblock everything
* do something with the information "stopped yes/no"

_Originally posted by @GitMensch in https://github.com/vim/vim/issues/9158#issuecomment-981981900_

**Comment by brammool** [comment]
I don't know what you mean with "end an instruction".  Are you talking about the communication with gdb?  That uses a gdb API.

**Comment by GitMensch** [comment]
Yes, we use the GDB API - but we don't do what the documentation says: when using async mode pass a token on each api-call and check it when reading responses to recognize the request it belongs to.

**Comment by GitMensch** [comment]
The GDB manuals had that in somewhere, I didn't found it with a quick ckeck.

But the [Input-Syntax](https://sourceware.org/gdb//onlinedocs/gdb/GDB_002fMI-Input-Syntax.html#GDB_002fMI-Input-Syntax) and [Output-Syntax](https://sourceware.org/gdb//onlinedocs/gdb/GDB_002fMI-Output-Syntax.html#GDB_002fMI-Output-Syntax) has the doc for the optional token - if we pass it in, we also get it in the response.

> If an input command was prefixed with a token then the corresponding output for that command will also be prefixed by that same token. 

Don't be confused like I was from the note that says that `async-output` has its tokens omitted - we commonly don't react on those, but on the other responses that are also returned in async mode.

**Comment by brammool** [comment]
> The GDB manuals had that in somewhere, I didn't found it with a quick ckeck.
> 
> But the
> [Input-Syntax](https://sourceware.org/gdb//onlinedocs/gdb/GDB_002fMI-Input-Syntax.html#GDB_002fMI-Input-Syntax)
> and
> [Output-Syntax](https://sourceware.org/gdb//onlinedocs/gdb/GDB_002fMI-Output-Syntax.html#GDB_002fMI-Output-Syntax)
> has the doc for the optional token - if we pass it in, we also get it
> in the response.

Thanks.  I wonder if all versions that we support have this
functionality.

It is unlikely I will find time for this, I hope someone can look into
this.  I haven't been very satisfied with how the communication with gdb
is done, thinking "there must be a better way".  What is called the
"machine interface" or MI looks too much like interactive commands.

-- 
hundred-and-one symptoms of being an internet addict:
76. Your ISP regards you as a business partner rather than as a customer.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by GitMensch** [comment]
> Thanks.  I wonder if all versions that we support have this functionality.

It is [documented for GDB 7.12](https://raw.githubusercontent.com/bminor/binutils-gdb/gdb-7.12-release/gdb/doc/gdb.texinfo) (search for `@code{@var{token}}`), so the answer is: yes.

> I haven't been very satisfied with how the communication with gdb is done, thinking "there must be a better way"

The "token" approach is actually much more "machine" than the current one - which I agree is more "interactive"+"guessing".

**Comment by laburnumT** [comment]
I think there is definitely a place for cleaning up some of the termdebug code. There are also examples like the assembly window using the human interaction command rather than the single line machine interface. And the possible move towards vim9script: https://github.com/vim/vim/pull/12403#discussion_r1299389600
